compiler/resolution/functionResolution.cpp

```diff
@@ -10624,15 +10624,8 @@ static LoopWithShadowVarsInterface*
   isForeachWhoseShadowVarsShouldBeResolved(SymExpr *se)
 {
   ForLoop* pfl = toForLoop(se->parentExpr);
-
-  // The pfl->shadowVariables().length > 0 part of the above condition is a
-  // bit of a hack to ensure we don't apply implicit intents on a loop where
-  // we haven't added an explicit intent (via a 'with' clause). Getting
-  // intents to work for 'foreach' loops is a bit of a work in progress and
-  // for the time being I would like to keep the behavior of loops that
-  // don't have a 'with' clause unchanged.
   if(pfl && pfl->isOrderIndependent() && se == pfl->indexGet() &&
-     (pfl->shadowVariables().length > 0))
+     !pfl->shouldExemptFromImplicitIntents())
```

Comment:
Not related to this PR or this change, but what's the purpose of `se` argument here, and that we are comparing it against the loop index?

Reply:
This function is called by `resolveExpr`. For `forall` loops we process them when we encounter a `symExpr` who's parent is the loop.

I want to do a similar thing for `foreach` loops, but I guess the way they're structured there's more than one `symExpr` who's parent is the loop, (so I need to add some filter that will hit exactly once for a given `foreach` loop). Anyway, one unique `symExpr` who's parent is the loop is `pfl->indexGet()`.

Reply:
I'll add this comment in above the function:

```
  // Note: this function really only makes sense in the context of resolveExpr.
  // It gets passed a symexpr, and we check to see if its parent is a foreach
  // loop. We only want to process the loop one time and since there are multiple
  // symexprs that are children of a foreach, we fix this to only return true when
  // we're at the symexpr for the index variable.
```

Which is still somewhat unsatisfying; I want to ask why do we need to process the `foreach` when we call `resolveExpr` with `expr` being this child symexpr instead of more directly when `expr` is the `foreach` itself? And I'm not sure what the answer is there other than: "that's what we do for `forall` loops".